src/libPMacc/test/particles/IdProvider.hpp

```diff
@@ -0,0 +1,133 @@
+/**
+ * Copyright 2016-2016 Alexander Grund
+ *
+ * This file is part of libPMacc.
+ *
+ * libPMacc is free software: you can redistribute it and/or modify
+ * it under the terms of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * libPMacc is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with libPMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <pmacc_types.hpp>
+#include <particles/IdProvider.hpp>
+#include <memory/buffers/HostDeviceBuffer.hpp>
+#include <eventSystem/EventSystem.hpp>
+
+#include <boost/mpl/list.hpp>
+#include <boost/mpl/for_each.hpp>
+#include <boost/mpl/int.hpp>
+#include <boost/test/unit_test.hpp>
+#include <set>
+#include <algorithm>
+#include <stdint.h>
+
+BOOST_AUTO_TEST_SUITE( particles )
+
+namespace bmpl = boost::mpl;
+
+namespace
+{
+    template<class T_IdProvider, class T_Box>
+    __global__ void generateIds(T_Box outputbox, uint32_t numThreads, uint32_t numIdsPerThread)
+    {
+        const uint32_t localId = blockIdx.x * blockDim.x + threadIdx.x;
+        if(localId >= numThreads)
+            return;
+        for(uint32_t i=0; i<numIdsPerThread; i++)
+            outputbox(i * numThreads + localId) = T_IdProvider::getNewId();
+    }
+}
+
+/** Boost.Test compatible function that checks if a value is in a collection
+ *  An error is returned, if the value is not found and shouldFind is true or
+ *  if the value is found and shouldFind is false
+ *  Use like: BOOST_REQUIRE(checkDuplicate(col, value, true|false));
+ */
+template<class T_Collection, typename T>
+boost::test_tools::predicate_result
+checkDuplicate(const T_Collection& col, const T& value, bool shouldFind)
```

Comment:
Is the parameter `shouldFind` useful? Or is it not more intuitive for the user to negate the result.

```C++
BOOST_REQUIRE(checkDuplicate(col, value));
BOOST_REQUIRE(!checkDuplicate(col, value));
```